pkg/manifests/manifests.go

```diff
@@ -0,0 +1,193 @@
+	}
+	return manifest, nil
+}
+
+func findImagesInManifest(data interface{}) ([]string, error) {
+	imageSet := make(map[string]struct{})
+
+	var findImages func(data interface{})
+	findImages = func(data interface{}) {
+		switch t := data.(type) {
+		case map[string]interface{}:
+			for k, v := range t {
+				if k == "image" {
+					if imageName, ok := v.(string); ok {
+						imageSet[imageName] = struct{}{}
+					}
+				}
+				findImages(v)
+			}
+		case []interface{}:
+			for _, v := range t {
+				findImages(v)
+			}
+		}
+	}
+
+	findImages(data)
+
+	images := make([]string, 0, len(imageSet))
+	for uniqueImage := range imageSet {
+		images = append(images, uniqueImage)
+	}
+
+	return images, nil
+}
+
+func copyManifests(src string, dest string) ([]string, error) {
+	if dest == "" {
+		return nil, fmt.Errorf("manifest destination directory not defined")
+	}
+
+	var manifestPaths []string
+
+	manifests, err := os.ReadDir(src)
+	if err != nil {
+		return nil, fmt.Errorf("reading manifest source dir: %w", err)
+	}
+
+	for _, manifest := range manifests {
+		manifestName := strings.ToLower(manifest.Name())
+		if filepath.Ext(manifestName) != ".yaml" && filepath.Ext(manifestName) != ".yml" {
+			zap.S().Warnf("Skipping %s as it is not a yaml file", manifest.Name())
+			continue
+		}
+
+		sourcePath := filepath.Join(src, manifest.Name())
+		destPath := filepath.Join(dest, manifest.Name())
+		err := fileio.CopyFile(sourcePath, destPath, fileio.NonExecutablePerms)
+		if err != nil {
+			return nil, fmt.Errorf("copying manifest file %s: %w", sourcePath, err)
+		}
+		manifestPaths = append(manifestPaths, destPath)
+
+	}
+
+	return manifestPaths, nil
+}
+
+func downloadManifests(ctx *image.Context, destPath string) ([]string, error) {
+	manifests := ctx.ImageDefinition.Kubernetes.Manifests.URLs
+	var manifestPaths []string
+
+	for index, manifestURL := range manifests {
+		filePath := filepath.Join(destPath, fmt.Sprintf("manifest-%d.yaml", index+1))
+		manifestPaths = append(manifestPaths, filePath)
+
+		if err := http.DownloadFile(context.Background(), manifestURL, filePath); err != nil {
+			return nil, fmt.Errorf("downloading manifest '%s': %w", manifestURL, err)
+		}
+	}
+
+	return manifestPaths, nil
+}
+
+func removeDuplicateImages(images []string) []string {
+	imagesSet := make(map[string]bool)
+	var formattedImages []string
+	for _, item := range images {
+		if _, value := imagesSet[item]; !value {
+			imagesSet[item] = true
+			formattedImages = append(formattedImages, item)
+		}
+	}
+
+	return formattedImages
+}
+
+func addImagesToRegistryDefinition(ctx *image.Context, manifestImages []string) {
+	for _, imageName := range manifestImages {
+		if !isInRegistry(ctx, imageName) {
+			containerImageDef := image.ContainerImage{
+				Name: imageName,
+			}
+			ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages = append(ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages, containerImageDef)
```

Comment:
This is the part which is the biggest problem, in my opinion. I don't think we should be modifying what the image definition itself contains - it's much better to have a slice which contains all of the `EmbeddedArtifactRegistry.ContainerImages` and then append the downloaded ones to that instead.